modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerMetricSource.java

```diff
@@ -0,0 +1,418 @@
+        }
+    }
+
+    void requestsActiveDecrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.requestsActive.decrement();
+        }
+    }
+
+    /**
+     * Gets processed requests.
+     *
+     * @return Processed requests.
+     */
+    public long requestsProcessed() {
+        Holder h = holder;
+
+        return h == null ? 0 : h.requestsProcessed.value();
+    }
+
+    void requestsProcessedIncrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.requestsProcessed.increment();
+        }
+    }
+
+    /**
+     * Gets failed requests.
+     *
+     * @return Failed requests.
+     */
+    public long requestsFailed() {
+        Holder h = holder;
+
+        return h == null ? 0 : h.requestsFailed.value();
+    }
+
+    void requestsFailedIncrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.requestsFailed.increment();
+        }
+    }
+
+    /**
+     * Gets active transactions.
+     *
+     * @return Active transactions.
+     */
+    public long transactionsActive() {
+        Holder h = holder;
+
+        return h == null ? 0 : h.transactionsActive.value();
+    }
+
+    /**
+     * Increments active transactions.
+     */
+    public void transactionsActiveIncrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.transactionsActive.increment();
+        }
+    }
+
+    /**
+     * Decrements active transactions.
+     */
+    public void transactionsActiveDecrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.transactionsActive.decrement();
+        }
+    }
+
+    /**
+     * Gets active cursors.
+     *
+     * @return Active cursors.
+     */
+    public long cursorsActive() {
+        Holder h = holder;
+
+        return h == null ? 0 : h.cursorsActive.value();
+    }
+
+    /**
+     * Increments active cursors.
+     */
+    public void cursorsActiveIncrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.cursorsActive.increment();
+        }
+    }
+
+    /**
+     * Decrements active cursors.
+     */
+    public void cursorsActiveDecrement() {
+        Holder h = holder;
+
+        if (h != null) {
+            h.cursorsActive.decrement();
+        }
+    }
+
+    @Override
+    public synchronized @Nullable MetricSet enable() {
+        Holder h = holder;
+
+        if (h == null) {
+            h = new Holder();
+            holder = h;
+        }
+
+        return h.metricSet;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public synchronized void disable() {
+        holder = null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public synchronized boolean enabled() {
+        return holder != null;
+    }
+
+    private static class Holder {
+        private final AtomicLongMetric connectionsInitiated =
+                new AtomicLongMetric("connections.Initiated", "Total initiated connections");
```

Comment:
I think we should either use `sessions` here or `connections` in other places as well.

Reply:
`connection != session`
* `connection`: someone connected to the endpoint, no data has been sent or received yet. This can even be a different application due to some network misconfiguration. It is useful to measure connections vs sessions to identify those misconfigurations.
* `session`: handshake successfully completed.

Reply:
OK, sounds reasonable.